src/ucc/api/ucc_def.h

```diff
@@ -66,7 +66,15 @@ typedef struct ucc_team*            ucc_team_h;
  * status of the collective operation, progress, or complete the collective
  * operation.
  */
-typedef struct ucc_coll_req*        ucc_coll_req_h;
+typedef struct ucc_coll_req* ucc_coll_req_h;
+typedef ucc_status_t (*ucc_coll_post_fn_t)(ucc_coll_req_h request);
+typedef ucc_status_t (*ucc_coll_finalize_fn_t)(ucc_coll_req_h request);
+typedef struct ucc_coll_req {
+    ucc_status_t           status;
+    ucc_coll_post_fn_t     post;
+    ucc_coll_finalize_fn_t finalize;
+} ucc_coll_req_t;
+
```

Comment:
why is it part of API? do you think user wants to access post and finalize functions?

Reply:
only to make implementations of ucc_collective_post and finalize static inlines and remove 1 extra fn call. We can this private and then move impl of ucc_collectve_post/finalize into ucc/core but we will have a fn call. Same logic applies to "status" field in the ucc_coll_req_t. If we make it part of API we can implement ucc_collective_test as just inlilne "status == UCC_OK" otherwise fn call again. Wdyt?

Reply:
the only issues I can think of with this approach is that we will need to track possible API changes, which means adding fields mask to coll_req, but that's probably fine

Reply:
I think we should not it not make it part of API and/or track it. Post/Finalize is a heavy weight operation, do you think a function call would make difference? In this case, maybe, pros doesn’t outweigh the cons.

Reply:
agree. moved those into internal coll_task. status is still part of public API - i think test is better to have w/o fn call - it is lightweight and called multiple times in a critical path.